esp_modem: CMUX mode improvements

david-cermak opened the pull request:
* Support for exit CMUX mode
* Support for CMUX mode in C-API

euripedesrocha commented on components/esp_modem/src/esp_modem_dce.cpp:
```diff
@@ -27,24 +29,50 @@ bool DCE_Mode::set(DTE *dte, ModuleIf *device, Netif &netif, modem_mode m)
+            }
+            netif.stop();
+            SignalGroup signal;
+            dte->set_read_cb([&](uint8_t *data, size_t len) -> bool {
```
I would suggest capture explicitly here, since it's a callback we may have lifetime issues if some references go out of scope. ASIO code uses that lifetime extension idea using `std::shared_from_this` to guarantee lifetime.

david-cermak replied:
The implicit capture was needed for logging as the `ESP_LOGx()` variants need statics and sometimes are defined and sometimes not (based on configured log severity).
I think we can capture the `signal` part only, hope it won't be necessary to use `shared_from_this`...

david-cermak replied:
captured only `signal`. Please resolve if you agree.

david-cermak replied:
> static don't need to be captured.

But `signal` is not static in this context, it's just an automatic variable. Or are you implying that we should make it static ?

david-cermak replied:
So I made the automatic variable a `shared_ptr<>` that's passed to the lambda as `weak_ptr<>`. This should address concerns related to local variable lifetime.

euripedesrocha commented on components/esp_modem/src/esp_modem_dte.cpp:
```diff
@@ -69,6 +69,23 @@ command_result DTE::command(const std::string &cmd, got_line_cb got_line, uint32
+    term = std::move(std::get<0>(ejected));
+    buffer = std::move(std::get<1>(ejected));
+    buffer_size = std::get<2>(ejected);
+    command_term = term.get(); // use command terminal as previously
```
About the lifetime control here. 
We have on `CMUX` usage of `std::unique_ptr` and here we gave up the automatic control to have a raw pointer.

david-cermak replied:
I see, this is a design flaw introduced by the CMUX mode. I could make it a `shared_ptr<>`, but that would mean promote every `unique_ptr<>` to the shared pointer, which I was a bit reluctant. I'll see if I can fix it...

david-cermak replied:
Refactored to used `shared_ptr<>` for both types of terminals, so they could be attached to `CMux` when switching to CMUX mode and detached when switching back.

david-cermak commented on components/esp_modem/examples/simple_cmux_client/main/simple_cmux_client_main.cpp:
```diff
@@ -87,7 +87,7 @@ extern "C" void app_main(void)
 
-    if (dce->set_mode(esp_modem::modem_mode::CMUX_MODE) && dce->set_mode(esp_modem::modem_mode::DATA_MODE)) {
+    if (dce->set_mode(esp_modem::modem_mode::CMUX_MODE)) {
         std::cout << "Modem has correctly entered multiplexed command/data mode" << std::endl;
```
Yes, it is still valid, as with the recent changes to the mode state machine, switching to CMUX creates two terminals behind the scene and sets the second one to the `data` mode.

david-cermak commented on components/esp_modem/include/cxx_include/esp_modem_cmux.hpp:
```diff
@@ -64,6 +64,13 @@ class CMux {
+     * @return  nullptr on failure
+     *          tuple of the original terminal and buffer on success
+     */
+    std::tuple<std::unique_ptr<Terminal>, std::unique_ptr<uint8_t[]>, size_t> deinit_and_eject();
```
I'm afraid we cannot make it an integral part of the `Terminal` (since it's shared with the CMUX-like terminal).

What we could do though is to bundle the `unique-ptr` and the size into a type, I'm on it.

david-cermak replied:
Created `unique_buffer` to hold the pointer, size and offset, and is `std::move`-able so we can simply transfer ownership between `CMux` and `DTE`

david-cermak commented on components/esp_modem/include/cxx_include/esp_modem_cmux.hpp:
```diff
@@ -84,6 +91,9 @@ class CMux {
     static uint8_t fcs_crc(const uint8_t frame[6]);     /*!< Utility to calculate FCS CRC */
     void data_available(uint8_t *data, size_t len);     /*!< Called when valid data available */
     void send_sabm(size_t i);                           /*!< Sending initial SABM */
+    void send_disc(size_t i);                           /*!< Sending closing request for each virtual terminal */
```
I'd rather keep the *disc* name here, as we already have a *close* command and they differ:
* `send_disc()` closes **one** virtual terminal
* `close()` exits the control mode of CMUX, so closes the entire mode.

Suggest renaming to `send_disconnect()`, WYT?

euripedesrocha replied:
Actually, `send_disconnect` was my first idea, but I saw the comment on sending closing. 

Maybe rename `close` to `exit_cmux`or `close_all`

david-cermak replied:
I wen with `send_disconnect()` and made a special case for closing the *control* terminal as `send_disconnect(0)` (since the control terminal is referred as terminal 0.

david-cermak commented on components/esp_modem/include/cxx_include/esp_modem_dce_module.hpp:
```diff
@@ -79,8 +79,17 @@ class GenericModule: public ModuleIf {
-            return set_command_mode() == command_result::OK;
+            Task::Delay(1000); // Mandatory 1s pause before
+            int retry = 0;
+            while (retry++ < 3) {
```
will add some comments, the point is that we're trying to `set_command_mode()`, but if that fails we might already be in the command mode, so the `sync()` just checks if we are.
That sequence is simply repeated (`retry` times) and whenever we succeed we call it a success and `return true;`

david-cermak commented:
> I've just tested this PR, but it fails to compile here:

Oh, sorry. I'm currently in the process of addressing the review comments, so it's very WIP now. If you like to test it, just drop the latest commit (9d51d61585727123bcd7f4233dee65304a4dff82 : have accepted some re-naming suggestions that break the build). This should work okay, although there are still some architectural *hacks* (like up-casting the DTE terminal to cmux-terminal, and some ptr ownership issues...) which should be cleaned up.

bbinet commented:
Ok, thanks.

david-cermak commented on components/esp_modem/include/cxx_include/esp_modem_cmux.hpp:
```diff
@@ -147,6 +157,7 @@ class CMuxInstance: public Terminal {
     }
     void start() override { }
     void stop() override { }
+    CMux* get_cmux() { return cmux.get(); }
```
Removed

david-cermak commented on components/esp_modem/include/cxx_include/esp_modem_buffer.hpp:
```diff
@@ -0,0 +1,50 @@
+ */
+struct unique_buffer {
+    explicit unique_buffer(size_t size);
+    explicit unique_buffer():
```
Removed, thanks!

> What a zero sized unique_buffer means in this context?

The reason was to be able to mark the buffer invalid, i.e. the same reason why e.g. `unique_ptr<>` could hold `nullptr`

hqs1 commented:
Hello, I found that if the AT command is continuously executed, the network request will easily time out, which will cause the network data to not be received normally. My idea is it possible to prioritize network requests over AT commands?
